Cap segment artifacts on very long running transactions (#3615)
newrelic/newrelic-ruby-agent: CHANGELOG.md

## dev

- **Feature: Add transaction_tracer.cap_segment_artifacts configuration option**

>   Long-running transactions with many segments can cause continuously increasing memory usage for the lifetime of the transaction. The agent now offers an opt-in `transaction_tracer.cap_segment_artifacts` configuration option (default `false`, local-only). When enabled, once [`transaction_tracer.limit_segments`](https://docs.newrelic.com/docs/apm/agents/ruby-agent/configuration/ruby-agent-configuration/#transaction_tracer-limit_segments) is reached, the agent also stops recording exclusive time for any segments created afterward in that transaction, reducing memory usage at the cost of less accurate timing data for the transaction.[PR#3615](https://github.com/newrelic/newrelic-ruby-agent/pull/3615)

Reviewer: What do you mean by `local-only`?

Author: I meant to remove that part (bc its weird claude stuff) when i was editing it, but i must have accidentally pressed undo one too many times at some point lol. I'll fix that. (0ef2b0e7fe3cd8d5eee955189150721225a03a76)

>   Long-running transactions with many segments can cause continuously increasing memory usage for the lifetime of the transaction. The agent now offers an opt-in `transaction_tracer.cap_segment_artifacts` configuration option (default `false`, local-only). When enabled, once [`transaction_tracer.limit_segments`](https://docs.newrelic.com/docs/apm/agents/ruby-agent/configuration/ruby-agent-configuration/#transaction_tracer-limit_segments) is reached, the agent also stops recording exclusive time for any segments created afterward in that transaction, reducing memory usage at the cost of less accurate timing data for the transaction.[PR#3615](https://github.com/newrelic/newrelic-ruby-agent/pull/3615)

Author: ```suggestion
  Long-running transactions with many segments can cause continuously increasing memory usage for the lifetime of the transaction. The agent now offers an opt-in `transaction_tracer.cap_segment_artifacts` configuration option (defaults to `false`). When enabled, once [`transaction_tracer.limit_segments`](https://docs.newrelic.com/docs/apm/agents/ruby-agent/configuration/ruby-agent-configuration/#transaction_tracer-limit_segments) is reached, the agent also stops recording exclusive time for any segments created afterward in that transaction, reducing memory usage at the cost of less accurate timing data for the transaction.[PR#3615](https://github.com/newrelic/newrelic-ruby-agent/pull/3615)
```

  The agent now detects Cloud Run and reports the GCP instance id as the hostname so individual instances can be distinguished. Before this change, all Google Cloud Run hostnames were `localhost`. This feature is controlled by the new `utilization.gcp_cloud_run.use_instance_as_host` configuration option (default `true`). Set `utilization.gcp_cloud_run.include_revision_in_host` (default `false`) to `true` to report the hostname as `{K_REVISION}-{instance id}` instead, where [`K_REVISION`](https://docs.cloud.google.com/run/docs/container-contract#env-vars) is the Cloud Run revision name. Thank you to [@kawa-onushi](https://github.com/kawa-onushi) for suggesting this improvement. [Issue#3295](https://github.com/newrelic/newrelic-ruby-agent/issues/3295) [PR#3609](https://github.com/newrelic/newrelic-ruby-agent/pull/3609/)

> - **Bugfix: Slow SQL recorded after transaction_tracer.limit_segments exceeded**

Reviewer: I think I understand that it's fixing the statement after the colon, but maybe being more explicit here would help? 

```suggestion
- **Bugfix: Slow SQL no longer recorded after transaction_tracer.limit_segments exceeded**
```
